src/Estimators/WalkerLogBuffer.h

```diff
@@ -0,0 +1,376 @@
+
+
+#ifndef QMCPLUSPLUS_WALKERLOGBUFFER_H
+#define QMCPLUSPLUS_WALKERLOGBUFFER_H
+
+
+#include <Configuration.h>
+#include "OhmmsPETE/OhmmsArray.h"
+#include "hdf/hdf_archive.h"
+
+#include <unordered_set>
+
+
+
+namespace qmcplusplus
+{
+
+
+/// Basic data types for walker log data
+struct WLog
+{
+  using Int  = long;                  // integer type
+  using Real = OHMMS_PRECISION_FULL;  // real type
+  using Comp = std::complex<Real>;    // complex type
+#ifndef QMC_COMPLEX
+  using PsiVal = Real;                // wavefunction type
+#else
+  using PsiVal = Comp;                // wavefunction type
+#endif
+};
+
+
+/** Record for an individual walker quantity being logd.
+ *
+ *    Helper struct for WalkerLogBuffer.
+ */
+struct WalkerQuantityInfo
+{
+  /// quantity name
+  std::string name;                
+  /// support up to 4D array quantity
+  enum{D0=0,D1,D2,D3,DMAX};        
+  /// array dimension
+  size_t dimension;                
+  /// total size
+  size_t size;                     
+  /// size of 1 unit of data
+  size_t unit_size;                
+  /// array shape
+  TinyVector<size_t, DMAX> shape;  
+  /// starting row index in buffer
+  size_t buffer_start;             
+  /// end range in buffer row
+  size_t buffer_end;               
+
+  WalkerQuantityInfo(const std::string& name_,size_t unit_size_,size_t buffer_start_,size_t n1=1,size_t n2=0,size_t n3=0,size_t n4=0)
+  {
+    name         = name_;
+    unit_size    = unit_size_;
+    buffer_start = buffer_start_;
+    shape[D0]    = n1;
+    shape[D1]    = n2;
+    shape[D2]    = n3;
+    shape[D3]    = n4;
+
+    dimension = 0;
+    size      = 1;
+    for (size_t d=0;d<DMAX;++d)
+      if (shape[d]>0)
+      {
+        dimension++;
+        size *= shape[d];
+      }
+    buffer_end = buffer_start + size*unit_size;
+  }
+};
+
+
+
+/** Data buffer for walker log quantities.
+ *
+ *    Each row in the buffer contains all quantities for one walker from a single step.
+ *    Rows are added throughout an MC block.
+ *    See WalkerLogCollector::collect()
+ *
+ *    Buffer data is written to HDF at the end of each MC block.
+ *    See WalkerLogManager::writeBuffers()
+ */
+template<typename T>
+class WalkerLogBuffer
+{
+public:
+  /// label for this data in HDF file
+  std::string label;         
+  /// marks first WalkerLogCollector::collect() call
+  bool first_collect;        
+  /// HDF file pointer
+  hsize_t hdf_file_pointer;  
+
+private:
+  /// index of current quantity during WalkerLogCollector::collect()
+  size_t quantity_pointer;   
+  /** buffer row location data for each walker quantity
+   *    used to populate "data_layout" field in HDF file
+   */
+  std::vector<WalkerQuantityInfo> quantity_info;
+  /// total size of walker data stored in a buffer row
+  size_t walker_data_size;
+  /// the walker data buffer itself
+  Array<T, 2> buffer;
+  /// array dimensions used in HDF file write
+  hsize_t dims[2];
```

Comment:
Seems being used only locally. So please keep it to the function writeHDF.

Reply:
Done